add TAMP register block for g0, g4, l5, u5 and wl

oll3 opened the pull request:
Adds TAMP register block for the devices where I could understand how it was mapped. Some devices seems to have their TAMP/backup/RTC more intertwined and not as a separate register block. Those are not included.

Also not really sure why only u5 and l5 was mapped automatically while I had to add them manually for g0, g4 and wl through extra/families. I guess some field is missing in one of the sources?

Mainly adding this for gaining access to backup registers. Tested on stm32wle5cc.

xoviat commented:
Have you looked in the source files to see why chips.rs is not picking it up?


https://embassy.dev/book/dev/developer_stm32.html

oll3 commented:
Seems `InstanceName="TAMP"` is not present in many of the xmls where the corresponding `TAMP` block is present in the svd.
Eg `stm32wle5.svd` has a `TAMP` peripheral but no  `InstanceName="TAMP"`  in `STM32WLE5CCUx.xml`.

How should this be handled?

xoviat commented on data/registers/tamp_g0.yaml:
```diff
@@ -0,0 +1,337 @@
+    - name: ITAMP6E
+      description: ITAMP6E
+      bit_offset: 21
+      bit_size: 1
```
consolidate to an array

xoviat replied:
consolidate to an array

xoviat commented on data/registers/tamp_g0.yaml:
```diff
@@ -0,0 +1,337 @@
+    - name: TAMP2MSK
+      description: TAMP2MSK
+      bit_offset: 17
+      bit_size: 1
```
consolidate to an array

xoviat commented on data/registers/tamp_g0.yaml:
```diff
@@ -0,0 +1,337 @@
+    - name: TAMP2NOER
+      description: TAMP2NOER
+      bit_offset: 1
+      bit_size: 1
```
consolidate to an array

xoviat commented on data/registers/tamp_g0.yaml:
```diff
@@ -0,0 +1,337 @@
+    - name: TAMP2TRG
+      description: TAMP2TRG
+      bit_offset: 25
+      bit_size: 1
```
consolidate to an array

xoviat commented on data/registers/tamp_g0.yaml:
```diff
@@ -0,0 +1,337 @@
+    - name: ITAMP6IE
+      description: ITAMP6IE
+      bit_offset: 21
+      bit_size: 1
```
consolidate to an array

xoviat commented on data/registers/tamp_g0.yaml:
```diff
@@ -0,0 +1,337 @@
+    - name: ITAMP6MF
+      description: ITAMP6MF
+      bit_offset: 21
+      bit_size: 1
```
consolidate to an array

xoviat commented on data/registers/tamp_g0.yaml:
```diff
@@ -0,0 +1,337 @@
+    - name: CITAMP7F
+      description: CITAMP7F
+      bit_offset: 22
+      bit_size: 1
```
consolidate to an array

xoviat commented on data/registers/tamp_g0.yaml:
```diff
@@ -0,0 +1,337 @@
+    - name: ITAMP7F
+      description: ITAMP7F
+      bit_offset: 22
+      bit_size: 1
```
consolidate to an array

xoviat reviewed (changes requested):
any fields, etc that have a number suffix need to be consolidated to an array. see the hrtim_v1.yaml for examples of this.

xoviat commented:
> How should this be handled?

If it's not in the XML, you've done the right thing. I will merge after you arrayize the fields, per the review.

xoviat commented:
> I will merge after you arrayize the fields, per the review.

Also, to be clear I mean all of the files. Not just the one I reviewed.

oll3 commented:
> any fields, etc that have a number suffix need to be consolidated to an array. see the hrtim_v1.yaml for examples of this.

I see, though many of these fields have flags which are not fully sequential. Should they be merged to arrays anyway?

Eg. ITAMPxIE from `tamp_wl.yaml` has flags on bits 18, 20, 21, 23:
```yaml
    - name: ITAMP3IE
      description: Internal tamper 3 interrupt enable
      bit_offset: 18
      bit_size: 1
    - name: ITAMP5IE
      description: Internal tamper 5 interrupt enable
      bit_offset: 20
      bit_size: 1
    - name: ITAMP6IE
      description: Internal tamper 6 interrupt enable
      bit_offset: 21
      bit_size: 1
    - name: ITAMP8IE
      description: Internal tamper 8 interrupt enable
      bit_offset: 23
      bit_size: 1
```

As an array I guess it would look like this, ignoring that offset 19 and 22 are not valid:
```yaml
    - name: ITAMPIE
      description: Internal tamper 3 interrupt enable
      bit_offset: 18
      bit_size: 1
      array:
        len: 6
        stride: 1
```

xoviat commented:
Use offsets in that case. Again, see hrtim for an example.

Dirbaio commented:
In this case the bit offsets line up nicely with the numbers, it's just that there's "holes" because some bits are unimplemented. In this case, just use a normal array. it's OK if you can write to the unimplemented bits, these are typically ignored anyway.

only use offsets if the bits really don't line up. For example if there was this:
- ITAMP1IE on bit 8
- ITAMP2IE on bit 9
- ITAMP3IE on bit 16
- ITAMP4IE on bit 17

oll3 commented:
> In this case the bit offsets line up nicely with the numbers, it's just that there's "holes" because some bits are unimplemented. In this case, just use a normal array. it's OK if you can write to the unimplemented bits, these are typically ignored anyway.
> 
> (using offsets would prevent writing to unimplemented bits, but then it no longer holds that TAMPn is at index n-1, which is confusing)
> ...

Makes sense!

I have made repeated fields into arrays and also removed some more 1 bit enums which seemed fairly obvious. And renamed some enums with confusing names (mainly for u5).

Thank you both for taking the time to review and the guidance, really appreciate it! :star2:

xoviat commented:
bender run
